File: app/src/main/java/com/qwy/designpattern/lagou/singletonpattern/java/SingletonPattern03.java
package com.qwy.designpattern.lagou.singletonpattern.java;


/**
 * 饿汉式
 */

public class SingletonPattern03 {
    // 这句就是饿汉模式的核心
    private static SingletonPattern03 SINGLETONPATTERN03 = new SingletonPattern03();

    public SingletonPattern03() {
    }


    public static SingletonPattern03 getInstance() {
        return SINGLETONPATTERN03;
    }
}

File: app/src/main/java/com/qwy/designpattern/lagou/singletonpattern/kotlin/SingletonPattern03.kt
package com.qwy.designpattern.lagou.singletonpattern.kotlin

/**
 * 线程安全的懒汉式
 */
class SingletonPattern03 private constructor() {


    companion object {
        private var instance: SingletonPattern03? = null
            get() {
                if (field == null) {
                    field = SingletonPattern03()
                }
                return field
            }

        @Synchronized
        fun get(): SingletonPattern03 {
            return instance!!
        }


    }

}